test/python/transpiler/test_elide_permutations.py

```diff
@@ -0,0 +1,425 @@
+        qc.measure(2, 2)
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 0)
+        expected.barrier(0, 1, 2)
+        expected.measure(2, 0)
+        expected.measure(1, 1)
+        expected.measure(0, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_permutation_at_beginning(self):
+        """Test permutation in beginning of bell is elided."""
+        qc = QuantumCircuit(3, 3)
+        qc.append(PermutationGate([2, 1, 0]), [0, 1, 2])
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(2)
+        expected.cx(1, 0)
+        expected.barrier(0, 1, 2)
+        expected.measure(2, 0)
+        expected.measure(1, 1)
+        expected.measure(0, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_permutation_at_end(self):
+        """Test permutation at end of bell is elided."""
+        qc = QuantumCircuit(3, 3)
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+        qc.append(PermutationGate([2, 1, 0]), [0, 1, 2])
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 2)
+        expected.barrier(0, 1, 2)
+        expected.measure(0, 0)
+        expected.measure(1, 1)
+        expected.measure(2, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_swap_and_permutation(self):
+        """Test a combination of swap and permutation gates."""
+        qc = QuantumCircuit(3, 3)
+        qc.append(PermutationGate([2, 1, 0]), [0, 1, 2])
+        qc.swap(0, 2)
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 2)
+        expected.barrier(0, 1, 2)
+        expected.measure(0, 0)
+        expected.measure(1, 1)
+        expected.measure(2, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_permutation_before_measure(self):
+        """Test permutation before measure is elided."""
+        qc = QuantumCircuit(3, 3)
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.append(PermutationGate([1, 2, 0]), [0, 1, 2])
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 2)
+        expected.barrier(0, 1, 2)
+        expected.measure(1, 0)
+        expected.measure(2, 1)
+        expected.measure(0, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+
+class TestElidePermutationsInTranspileFlow(QiskitTestCase):
```

Comment:
Should we also test adding the pass in invalid positions (e.g. after layout), and having both it and `FinalizeLayout` run twice?

Reply:
I'll add a test for the after layout case. But did you have something in mind for the double case? It's a bit tricky in practice as the pass really isn't designed to be run more than once. The above todo you commented on in: https://github.com/Qiskit/qiskit/pull/9523#discussion_r1586599431 is for the case where you have `virtual_permutation_layout` already set and ensure the pass composes the layouts correctly. We could workaround that by calling `FinalizeLayouts` before the second execution, but in practice there isn't anything to eliminate a second time unless we insert more swaps after the first time

Reply:
Added the test case for after layout in: https://github.com/Qiskit/qiskit/pull/9523/commits/f5b6778e524ff5af540cf66fb0a42eb0a2a0b6c8